_deploy/r/gnoswap/gnft/gnft.gno

```diff
@@ -148,75 +197,121 @@ func Render(path string) string {
 	}
 }
 
-// Util
-func assertIsAdmin(address std.Address) {
-	if address != admin {
-		panic(addDetailToError(
-			errNoPermission,
-			ufmt.Sprintf("gnft.gno__assertIsAdmin() || only admin(%s) can call this function, called from %s", admin.String(), address.String()),
-		))
-	}
-}
+// SetTokenURI sets the metadata URI using a randomly generated SVG image
+// Parameters:
+//   - tid: The token ID to set the URI for
+//
+// Returns:
+//   - string: The token URI
+//   - error: Returns an error if the URI setting fails, nil otherwise
+func SetTokenURI(tid grc721.TokenID) (string, error) {
+	common.IsHalted()
 
-func Exists(tid grc721.TokenID) bool {
-	_, err := gnft.OwnerOf(tid)
+	tokenURI := genImageURI(generateRandInstance())
+	ok, err := gnft.SetTokenURI(tid, grc721.TokenURI(tokenURI))
 	if err != nil {
-		return false
+		return "", err
 	}
 
-	return true
+	prevAddr, prevPkgPath := getPrevAsString()
+	std.Emit(
+		"SetTokenURI",
+		"prevAddr", prevAddr,
+		"prevRealm", prevPkgPath,
+		"lpTokenId", string(tid),
+		"internal_tokenURI", tokenURI,
+	)
+
+	return tokenURI, nil
 }
 
-func SetTokenURI(tid grc721.TokenID) {
+// SetTokenURILast sets metadata URI for last minted nft token
+// Returns:
+//   - error: Returns an error if the URI setting fails, nil otherwise
+func SetTokenURILast() error {
 	common.IsHalted()
 
-	// rand instance
-	seed1 := uint64(time.Now().Unix() + int64(TotalSupply()))
-	seed2 := uint64(time.Now().UnixNano() + int64(TotalSupply()))
-	pcg := rand.NewPCG(seed1, seed2)
-	r := rand.New(pcg)
-
-	tokenURI := genImageURI(r)
-	ok, _ := gnft.SetTokenURI(tid, grc721.TokenURI(tokenURI))
-	if ok {
-		prevAddr, prevRealm := getPrev()
-		std.Emit(
-			"SetTokenURI",
-			"prevAddr", prevAddr,
-			"prevRealm", prevRealm,
-			"lpTokenId", string(tid),
-			"internal_tokenURI", tokenURI,
-		)
+	caller := std.PrevRealm().Addr()
+	tokenListByCaller, _ := getTokenList(caller)
+	lenTokenListByCaller := len(tokenListByCaller)
+	if lenTokenListByCaller == 0 {
+		return nil
```

Comment:
When calling mint and SetTokenURILast with a Multi message, to prevent abnormal behavior, panic should be handled if the caller does not have a tokenList. It should not terminate normally with a nil return. The transaction will be treated as successful.

Reply:
76a094d99f03c6480b3d3668d1ac1c52db1667ee